### PR TITLE
WIP: util: upload OOM record file to S3

### LINKED ISSUE #37409: record more information when TiDB is about to occur OOM
## Enhancement
record more information：

1. topN-mem-consumer
- current explain analyze
- desensitized SQL text
2. memory related variables
- system variables
- session variables

### PR DESCRIPTION
**Pull request by mengxin9014**
### What problem does this PR solve?

Issue Number: close #37409

Problem Summary:
record more information when TiDB is about to occur OOM and write the information to S3

### What is changed and how it works?
record more information：
1. topN-mem-consumer
- current explain analyze
- desensitized SQL text
2. memory related variables
- system variables
- session variables
3. TopN memory consumed processes at the OS level

If parameter is configured, we upload all OOM alarm information to specified S3

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
record more OOM alarm information and upload it to S3.
```

### CODE REVIEW

File: util/expensivequery/memory_usage_alarm.go [hunk 1]
```diff
@@ -44,9 +59,66 @@ type memoryUsageAlarm struct {
...
+			S3ForcePathStyle: aws.Bool(record.s3Conf.s3ForcePathStyle),
+		})
+		if err != nil {
+			logutil.BgLogger().Error("create s3 new session fail", zap.Error(err))
+			return
+		}
+		uploader := s3manager.NewUploader(sess)
+
+		for _, filename := range filenames {
+			file, err := os.OpenFile(filename, os.O_RDONLY, 0600)
```
Comment (hawkingrei) [inline util/expensivequery/memory_usage_alarm.go:94]: defer in the loop is bad. I think you can refactor the code block in the loop into a function.

Comment (mengxin9014) [reply util/expensivequery/memory_usage_alarm.go]: > defer in the loop is bad. I think you can refactor the code block in the loop into a function.

done.

File: util/expensivequery/memory_usage_alarm.go [hunk 2]
```diff
@@ -17,6 +17,7 @@ package expensivequery
 import (
 	"context"
 	"fmt"
+	"go.uber.org/zap/zapcore"
```
Comment (hawkingrei) [inline util/expensivequery/memory_usage_alarm.go:20]: It is disorder.

Comment (mengxin9014) [reply util/expensivequery/memory_usage_alarm.go]: > It is disorder.

done

**Pull request closed without merging**